In [6]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression, SGDRegressor
from sklearn.datasets import load_boston, load_iris
from sklearn.preprocessing import StandardScaler ,MinMaxScaler, RobustScaler
from sklearn.model_selection import train_test_split, cross_val_score 
from sklearn.pipeline import make_pipeline
from sklearn.datasets import fetch_california_housing
from statsmodels.stats.outliers_influence import variance_inflation_factor

import matplotlib.pyplot as plt
import matplotlib
matplotlib.rcParams['font.family']='Malgun Gothic'
matplotlib.rcParams['axes.unicode_minus'] = False

1. trees.csv를 읽어들여서 아래에 대해  
1) 상관관계, 다중공선성을 확인하시요  
2) crossvalidation score 를 구하시요.  
3) 아래의 데이터를 이용하여 Volume을 예측하시요  
Girth 8.8, 10.5
Height 63, 72

In [3]:
tree_df = pd.read_csv('data/trees.csv')
tree_df.head()

Girth  Height  Volume
0    8.3      70    10.3
1    8.6      65    10.3
2    8.8      63    10.2
3   10.5      72    16.4
4   10.7      81    18.8

In [4]:
# feature와 label로 분리
x_data = tree_df.iloc[:,:-1]
y_data = tree_df.iloc[:,-1]

In [5]:
# 학습데이터:테스트데이터 = 8:2
x_train, x_test, y_train, y_test = train_test_split(x_data, y_data.values, test_size = 0.2, random_state=1) # ndarray

In [7]:
# 1) 상관관계, 다중공선성을 확인하시오.
display(tree_df.corr())

vif = pd.DataFrame()
vif['VIF Factor']= [ variance_inflation_factor(tree_df.values, i )  for i in range( tree_df.shape[1] ) ]
vif['features'] = tree_df.columns
display(vif)


Girth   Height    Volume
Girth   1.000000  0.51928  0.967119
Height  0.519280  1.00000  0.598250
Volume  0.967119  0.59825  1.000000

VIF Factor features
0  158.198657    Girth
1   62.939266   Height
2   32.993875   Volume

In [8]:
# 2) crossvalidation score 를 구하시오.
model = make_pipeline(StandardScaler(),SGDRegressor(verbose=1))
model.fit(x_train, y_train)
r2scores = cross_val_score(model, x_data, y_data, cv=5, scoring='r2',verbose=1)
print(r2scores)
print(r2scores.mean())

-- Epoch 1
Norm: 2.44, NNZs: 2, Bias: 3.692865, T: 24, Avg. loss: 494.715027
Total training time: 0.00 seconds.
-- Epoch 2
Norm: 3.97, NNZs: 2, Bias: 6.091094, T: 48, Avg. loss: 384.242360
Total training time: 0.00 seconds.
-- Epoch 3
Norm: 5.08, NNZs: 2, Bias: 7.993495, T: 72, Avg. loss: 314.911363
Total training time: 0.00 seconds.
-- Epoch 4
Norm: 5.99, NNZs: 2, Bias: 9.606578, T: 96, Avg. loss: 264.998751
Total training time: 0.00 seconds.
-- Epoch 5
Norm: 6.75, NNZs: 2, Bias: 11.001555, T: 120, Avg. loss: 226.038400
Total training time: 0.00 seconds.
-- Epoch 6
Norm: 7.39, NNZs: 2, Bias: 12.237904, T: 144, Avg. loss: 195.217341
Total training time: 0.00 seconds.
-- Epoch 7
Norm: 7.95, NNZs: 2, Bias: 13.339909, T: 168, Avg. loss: 169.996989
Total training time: 0.00 seconds.
-- Epoch 8
Norm: 8.44, NNZs: 2, Bias: 14.333211, T: 192, Avg. loss: 149.165138
Total training time: 0.00 seconds.
-- Epoch 9
Norm: 8.87, NNZs: 2, Bias: 15.239104, T: 216, Avg. loss: 131.738070
Total training ti

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.



-- Epoch 7
Norm: 3.96, NNZs: 2, Bias: 11.066167, T: 175, Avg. loss: 94.541119
Total training time: 0.03 seconds.
-- Epoch 8
Norm: 4.23, NNZs: 2, Bias: 11.878239, T: 200, Avg. loss: 82.599630
Total training time: 0.03 seconds.
-- Epoch 9
Norm: 4.47, NNZs: 2, Bias: 12.614812, T: 225, Avg. loss: 72.564717
Total training time: 0.03 seconds.
-- Epoch 10
Norm: 4.69, NNZs: 2, Bias: 13.285756, T: 250, Avg. loss: 64.037359
Total training time: 0.03 seconds.
-- Epoch 11
Norm: 4.88, NNZs: 2, Bias: 13.900855, T: 275, Avg. loss: 56.741660
Total training time: 0.03 seconds.
-- Epoch 12
Norm: 5.06, NNZs: 2, Bias: 14.464935, T: 300, Avg. loss: 50.466131
Total training time: 0.03 seconds.
-- Epoch 13
Norm: 5.23, NNZs: 2, Bias: 14.985822, T: 325, Avg. loss: 45.040651
Total training time: 0.03 seconds.
-- Epoch 14
Norm: 5.38, NNZs: 2, Bias: 15.467600, T: 350, Avg. loss: 40.325809
Total training time: 0.03 seconds.
-- Epoch 15
Norm: 5.51, NNZs: 2, Bias: 15.913640, T: 375, Avg. loss: 36.214923
Total train

[Parallel(n_jobs=1)]: Done   5 out of   5 | elapsed:    0.1s finished


In [9]:
# 3) 아래의 데이터를 이용하여 Volume을 예측하시오.
model.predict([[8.8, 63],[10.5,72]])

array([ 5.6129294 , 16.28177408])

2. volume이 40 이상이면 크다
30이상이면 보통 미만이면 적음으로
아래와 같이 출력하시요

|volume | 정도|
|-----|-----|
 |10.3| 적음|

In [13]:
volume_list = tree_df['Volume'].values
print(volume_list)
degree = []

for i in volume_list:
    if i >= 40:
        degree.append('크다')
    elif i>=30:
        degree.append('보통')
    else:
        degree.append('적음')

dic = {'volume':volume_list,
      'degree':degree}        
df = pd.DataFrame(dic)
df.head()
    

[10.3 10.3 10.2 16.4 18.8 19.7 15.6 18.2 22.6 19.9 24.2 21.  21.4 21.3
 19.1 22.2 33.8 27.4 25.7 24.9 34.5 31.7 36.3 38.3 42.6 55.4 55.7 58.3
 51.5 51.  77. ]


volume degree
0    10.3     적음
1    10.3     적음
2    10.2     적음
3    16.4     적음
4    18.8     적음

3. Height  가 가장 작은값과 큰값을 구하시오.

In [14]:
display(tree_df['Height'].max())
display(tree_df['Height'].min())

87

63

4. girth(테두리) 가 가장큰 top5를 구하시오(girth, height, volume)

In [20]:
tree_df.sort_values(by='Girth',axis=0,ascending=False,inplace=True)
tree_df.head()

Girth  Height  Volume
30   20.6      87    77.0
29   18.0      80    51.0
28   18.0      80    51.5
27   17.9      80    58.3
26   17.5      82    55.7